sdk/lib/udmihelp/udmihelp.c

```diff
@@ -0,0 +1,169 @@
+    if (!hKernel)
+        return NULL;
+
+    // try use GetSystemFirmwareTable (works on NT 5.2 and higher)
+    pGetSystemFirmwareTable = (void *)GetProcAddress(hKernel, "GetSystemFirmwareTable");
+    if (pGetSystemFirmwareTable)
+    {
+        dwBytes = pGetSystemFirmwareTable('RSMB', 0, NULL, 0);
+        if (dwBytes > 0)
+        {
+            pBuffer = malloc(dwBytes);
+            if (!pBuffer)
+            {
+                return NULL;
+            }
+            dwCheck = pGetSystemFirmwareTable('RSMB', 0, pBuffer, dwBytes);
+            if (dwCheck != dwBytes)
+            {
+                free(pBuffer);
+                return NULL;
+            }
+        }
+    }
+    if (dwBytes == 0)
+    {
+        // try use registry (works on NT 5.1)
+        if (RegOpenKeyW(HKEY_LOCAL_MACHINE,
+                        L"SYSTEM\\CurrentControlSet\\Services\\mssmbios\\Data",
+                        &hKey) != ERROR_SUCCESS)
+        {
+            return NULL;
+        }
+
+        if (RegQueryValueExW(hKey,
+                             L"SMBiosData",
+                             NULL,
+                             &dwType,
+                             NULL,
+                             &dwBytes) != ERROR_SUCCESS || dwType != REG_BINARY)
+        {
+            RegCloseKey(hKey);
+            return NULL;
+        }
+
+        pBuffer = malloc(dwBytes);
+        if (!pBuffer)
+        {
+            RegCloseKey(hKey);
+            return NULL;
+        }
+
+        if (RegQueryValueExW(hKey,
+                             L"SMBiosData",
+                             NULL,
+                             &dwType,
+                             pBuffer,
+                             &dwBytes) != ERROR_SUCCESS || dwType != REG_BINARY)
+        {
+            free(pBuffer);
+            RegCloseKey(hKey);
+            return NULL;
+        }
+
+        RegCloseKey(hKey);
+    }
+    ParseSMBiosTables(pBuffer, dwBytes, Strings);
+    return pBuffer;
+}
+
+/* Trim converted DMI string */
+VOID
+TrimDmiStringW(
+    _Inout_ LPWSTR lpStr)
+{
+    unsigned Length;
+    int i = 0;
+
+    if (!lpStr)
+        return;
+
+    Length = wcslen(lpStr);
+    if (Length == 0)
+        return;
+
+    /* Trim leading spaces */
+    while (i < Length && lpStr[i] <= 32)
+    {
+        i++;
+    }
+    if (i < Length)
+    {
+        if (i > 0)
+        {
+            Length -= i;
+            memmove(lpStr, lpStr + i, (Length + 1) * sizeof(WCHAR));
+        }
+
+        /* Trim trailing spaces */
+        while (Length && lpStr[Length-1] <= 32)
+        {
+            lpStr[Length-1] = L'\0';
+            --Length;
+        }
+    }
+    else
+    {
+        *lpStr = 0;
+    }
+}
+
+/* Convert string from SMBIOS */
+unsigned
+GetSMBiosStringW(
+    _In_ PCHAR DmiString,
+    _Out_ LPWSTR lpBuf,
+    _In_ DWORD cchBuf,
+    _In_ BOOL bTrim)
+{
+    unsigned cChars;
+
+    if (!DmiString)
+        return 0;
+
+    cChars = strlen(DmiString);
```

Comment:
You're assuming the number of chars in ansi and unicode strings will be equal. It's better to use MultiByteToWideChar's return value instead.